hydromt/driver/base_driver.py

```diff
@@ -0,0 +1,94 @@
+"""Base class for different drivers."""
+
+from abc import ABC
+from typing import Any, Callable, ClassVar
+
+from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator
+
+from hydromt.metadata_resolver import MetaDataResolver
+from hydromt.metadata_resolver.resolver_plugin import RESOLVERS
+
+
+class BaseDriver(BaseModel, ABC):
+    """Base class for different drivers.
+
+    Is used to implement common functionality.
+    """
+
+    name: ClassVar[str]
+    metadata_resolver: MetaDataResolver = Field(
+        default_factory=RESOLVERS.get("convention")
+    )
+    model_config = ConfigDict(extra="allow")
+
+    @field_validator("metadata_resolver", mode="before")
+    @classmethod
+    def _validate_metadata_resolver(cls, v: Any):
+        if isinstance(v, str):
+            if v not in RESOLVERS:
+                raise ValueError(f"unknown MetaDataResolver: '{v}'.")
+            return RESOLVERS.get(v)()
+        elif isinstance(v, dict):
+            try:
+                name: str = v.pop("name")
+                if name not in RESOLVERS:
+                    raise ValueError(f"unknown MetaDataResolver: '{name}'.")
+                else:
+                    return RESOLVERS.get(name).model_validate(v)
+            except KeyError:
+                # return default when name is missing
+                return cls.model_fields.get("metadata_resolver").default_factory(**v)
+
+        elif v is None:  # let default factory handle it
+            return None
+        elif hasattr(v, "resolve"):  # MetaDataResolver duck-typing
+            return v
+        else:
+            raise ValueError(
+                "metadata_resolver should be string, dict or MetaDataResolver."
+            )
+
+    @model_validator(mode="wrap")
+    @classmethod
+    def _init_driver(cls, data: Any, handler: Callable):
+        """Initialize the subclass based on the 'name' class variable.
+
+        All DataSources should be parsed based on their `name` class variable;
+        e.g. a dict with `name` = `pyogrio` should be parsed as a `PyogrioDriver`.
+        This class searches all subclasses until the correct driver is found and
+        initialized that one.
+        This allows an API as: `BaseDriver.model_validate(pyogrio_driver_dict)` or
+        `BaseDriver(pyogrio_driver_dict)`, but also `PyogrioDriver(pyogrio_driver_dict)`.
+
+        Inspired by: https://github.com/pydantic/pydantic/discussions/7008#discussioncomment
+        """
+        if not isinstance(data, dict):
+            # Other objects should already be the correct subclass.
+            return handler(data)
```

Comment:
I think you can just do a `before`, and only do transformations if needed, otherwise just return the value `data` that you received and let pydantic resolve it.

Reply:
Pydantic does not allow `before` here, as we would otherwise still run model validation for the `BaseDriver`, while we are actually running the validation for the subclass (e.g. `PyogrioDriver`). So after validating this one, we just want to exit without calling pydantic any further. All other validators are already called at `target_cls.model_validate(data)`.